### PR TITLE
Initial optimisation functionality

### PR DESCRIPTION
**Pull request by jotslo**
ASTs now automatically get optimised when compiling.

Example:
An AST derived from the following code...
```
i32 test = 5;
test = test + 1; //6

fn void unused_func
{
    test = 73;
    test = test - 1;
}

fn i32 main
{
    test = test + 1; //7

    if (true == true)
    {
        if (test < 1) //false
        {
            print "wrong";
        }
        else if ((test + 1) / 2 == 4) //true
        {
            print test; //7
        }
    }

    return 0;
}
```

will be optimised to...

```
fn i32 main
{
    {
        {
            // scopes remain in-case code relies on scope
            print 7;
        }
    }

    return 0;
}
```

Optimisations include:
- If/else/elseif statements with determinable values are replaced with their inner scopes where possible
- Unused variables/functions are removed from the AST
- Expressions containing determinable symbols/values are replaced with their result

I have tested all the edge cases that I could think of, and none of the unit tests appear to give incorrect ASTs, so hopefully this should be mergeable without any issues. Let me know if there are any issues though.

TODO:
- Functions, for loops and while loops declare variables as indeterminable after being called/used to prevent inaccurate optimisations. Will look at improving this functionality in the future.

### CODE REVIEW

**Comment by aaf6aa** [comment]
I have a question, why'd you remove .gitignore?

**Comment by jotslo** [comment]
> I have a question, why'd you remove .gitignore?

That’s my bad, I intended to have the pull request ignore the changes I made but accidentally removed the files - I’ve added them back to the branch

**Comment by aaf6aa** [comment]
Alright, I'll continue my review. Though, if it is possible, could you revert the commits instead of adding the files back, doing this destroys the file history. If not I'll do it myself.

**Comment by jotslo** [comment]
> Alright, I'll continue my review. Though, if it is possible, could you revert the commits instead of adding the files back, doing this destroys the file history. If not I'll do it myself.

If you’re looking through now and are happy to, please could you sort that then please? Otherwise, I can redo it in a few hours when I’ve got access to a PC - whichever is easiest

File: .gitignore [hunk 1]
```diff
@@ -49,3 +49,6 @@ generated/
 # vscode configs
 .vscode/settings.json
 .vscode/c_cpp_properties.json
+txt.txt
```
Comment (aaf6aa) [inline .gitignore:52]: Ignore these files with a wildcard, like "*.txt", instead of writing each filename individually

File: src/tin.c [hunk 2]
```diff
@@ -108,6 +109,7 @@ int main(int argc, char **argv)
 		print_step("Parsing and preprocessing failed\n");
 		goto fail;
 	}
+	optimize(mod, mod->ast_root);
```
Comment (aaf6aa) [inline src/tin.c]: The code currently cannot compile some example files, such as examples/include.tin, and instead segfaults. Omitting the call to optimize results in it compiling successfully, so there must be an issue within the optimization code.

File: src/optimisation.c [hunk 3]
```diff
@@ -1,78 +1,547 @@
...
+    this should be ideally improved to a state where function calls can be replaced entirely
+    with literal values, if that's possible with the given source code
+    */
+    int count = 0;
+    vector* children;
+
+    while (children != node->children && count < 10)
+    {
+        count++;
+        children = vector_copy(node->children);
```
Comment (aaf6aa) [inline src/optimisation.c:564]: Memory leak

File: src/optimisation.c [hunk 4]
```diff
@@ -1,78 +1,547 @@
...
+        if (determinable)
         {
-            ast_node* node1 = ast_get_child(condition, 0);
-            ast_node* node2 = ast_get_child(condition, 1);
-            
-            if (node1->type == AstIntegerLit && node2->type == AstIntegerLit)
+            if (is_int(variable->dtype) && variable->is_literal)
             {
-                int response = eval_condition(condition, node1, node2);
+                ast_node* new_node = ast_new(AstIntegerLit);
```
Comment (aaf6aa) [inline src/optimisation.c:310]: Memory leak

File: src/optimisation.c [hunk 5]
```diff
@@ -1,78 +1,547 @@
...
+                ast_node* data_type_node = ast_new(AstDataType);
```
Comment (aaf6aa) [inline src/optimisation.c:311]: Memory leak

File: src/optimisation.c [hunk 6]
```diff
@@ -1,78 +1,547 @@
...
+        }
+    }
+    return false;
+}
+
+void reclass_as_bool(ast_node* node, bool value)
 {
-    long long int value1 = node1->value.integer;
-    long long int value2 = node2->value.integer;
+    ast_node* new_node = ast_new(AstBoolLit);
```
Comment (aaf6aa) [inline src/optimisation.c:33]: Memory leak

File: src/optimisation.c [hunk 7]
```diff
@@ -1,78 +1,547 @@
...
+                ast_node* new_node = ast_new(AstIntegerLit);
+                ast_node* data_type_node = ast_new(AstDataType);
+                node->type = AstIntegerLit;
+                node->value.integer = variable->value.integer;
 
-                if (response != 0)
-                {
-                    return ast_get_child(node, response);
-                }
+                data_type_node->value.dtype = data_type_new(
```
Comment (aaf6aa) [inline src/optimisation.c:315]: Memory leak

**Review by aaf6aa — CHANGES_REQUESTED** [review]
You should make sure that all files in `examples/` compile correctly with your changes as they are basically our baseline until automated testing is implemented and should not be broken by changes, ignore files in `units/` and the units you wrote here though, they're not required to be runnable.

Additionally, clean up the memory leaks I pointed out. Use `make memcheck` and `make memcheck_full` to see where leaks are happening; these commands can also point out what's causing segmentation faults if you are getting them.

Also merge main into this branch so your checks are relevant, there have been some changes to the makefile jsyk

File: src/tin.c [hunk 8]
```diff
@@ -1,5 +1,6 @@
 #include "interpreter.h"
 #include "module.h"
+#include "optimisation.h"
```
Comment (aaf6aa) [inline src/tin.c:3]: The main function is called `optimize`, check your s's and z's.

**Review by aaf6aa — APPROVED** [review]

**Comment by aaf6aa** [comment]
well done, you've surprised me, i approve this PR

**Review by XiiiC — APPROVED** [review]
Looks good! great work!
